extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -149,8 +157,19 @@ private Object getValueAtTargetIndex(@Nullable Integer targetIndex,
             if (rowCells == null) {
                 throw new IndexOutOfBoundsException();
             }
-            row = new RowN(rowCells);
-            indexToRow.putIfAbsent(targetIndex, row);
+            row = indexToRow.get(targetIndex);
+            if (row == null) {
+                // RowN is a wrapper around already accounted cells.
+                // We need to account only for RowN shallow size which is done internally in ramBytesUsed().
```

Comment:
Looks like the wrapper is actually only needed to evaluate the RowInput/CollectExpressions below, this means we could share it:

```diff
diff --git a/extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java b/extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java
index 55aa545a493..0922dc0ee3d 100644
--- a/extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java
+++ b/extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java
@@ -27,15 +27,14 @@
 import java.util.function.LongConsumer;
 
 import org.jetbrains.annotations.Nullable;
 
 import com.carrotsearch.hppc.IntObjectHashMap;
-import com.carrotsearch.hppc.RamUsageEstimator;
 
 import io.crate.data.Input;
 import io.crate.data.Row;
-import io.crate.data.RowN;
+import io.crate.data.UnsafeArrayRow;
 import io.crate.execution.engine.collect.CollectExpression;
 import io.crate.execution.engine.window.WindowFrameState;
 import io.crate.execution.engine.window.WindowFunction;
 import io.crate.metadata.Functions;
 import io.crate.metadata.functions.BoundSignature;
@@ -147,34 +146,28 @@ private Object getValueAtTargetIndex(LongConsumer nonRowMemoryReserver,
                                          List<? extends CollectExpression<Row, ?>> expressions,
                                          Input<?> ... args) {
         if (targetIndex == -1) {
             throw new IndexOutOfBoundsException();
         }
-        Row row;
-        if (indexToRow.containsKey(targetIndex)) {
-            row = indexToRow.get(targetIndex);
-        } else {
-            Object[] rowCells = currentFrame.getRowInPartitionAtIndexOrNull(targetIndex);
-            if (rowCells == null) {
-                throw new IndexOutOfBoundsException();
-            }
-            row = indexToRow.get(targetIndex);
+        Object[] row = indexToRow.get(targetIndex);
+        if (row == null) {
+            row = currentFrame.getRowInPartitionAtIndexOrNull(targetIndex);
             if (row == null) {
-                // RowN is a wrapper around already accounted cells.
-                // We need to account only for RowN shallow size which is done internally in ramBytesUsed().
-                row = new RowN(rowCells);
-                long ramBytesPrev = indexToRow.ramBytesUsed();
-                indexToRow.put(targetIndex, row);
-                long deltaUsed = indexToRow.ramBytesUsed() - ramBytesPrev;
-                if (deltaUsed > 0) {
-                    // Due to object size alignment in the underlying shallowUsedSizeOfArray, delta can be zero.
-                    nonRowMemoryReserver.accept(RamUsageEstimator.alignObjectSize(deltaUsed));
-                }
+                throw new IndexOutOfBoundsException();
+            }
+            // cells+ data within are owned by the WindowFunctionBatchIterator/WindowFrameState and do not need accounting.
+            // But the indexToRow map overhead needs to be accounted for
+            long ramBytesPrev = indexToRow.ramBytesUsed();
+            indexToRow.put(targetIndex, row);
+            long deltaUsed = indexToRow.ramBytesUsed() - ramBytesPrev;
+            if (deltaUsed > 0) {
+                nonRowMemoryReserver.accept(deltaUsed);
             }
         }
+        sharedRow.cells(row);
         for (CollectExpression<Row, ?> expression : expressions) {
-            expression.setNextRow(row);
+            expression.setNextRow(sharedRow);
         }
         return args[0].value();
     }
 
     private int getTargetIndex(int idxInPartition, int offset) {
@@ -193,11 +186,12 @@ private int getTargetIndex(int idxInPartition, int offset) {
     private int resolvedDirection;
     /* cachedNonNullIndex and idxInPartition forms an offset window containing 'offset' number of non-null elements and any number of nulls.
        The main perf. benefits will be seen when series of nulls are present.
        In the cases where idxInPartition is near the bounds that the cache cannot point to the valid index, it will hold a null. */
     private int cachedNonNullIndex;
-    private IntObjectHashMap<Row> indexToRow;
+    private IntObjectHashMap<Object[]> indexToRow;
+    private UnsafeArrayRow sharedRow = new UnsafeArrayRow();
 
     private OffsetValueFunctions(Signature signature, BoundSignature boundSignature, int directionMultiplier) {
         this.signature = signature;
         this.boundSignature = boundSignature;
         this.directionMultiplier = directionMultiplier;
```

Reply:
Thanks, applied in https://github.com/crate/crate/pull/16054/commits/e8afd7a2bd5f582345fa9ac4f40d7d8a72047c3d